---
title: "Pooled Analysis"
author: "Aimee Taylor and James Watson"
output:
  html_document:
    df_print: paged
    keep_md: yes
  pdf_document: default
---

# Preamble

Load R packages, functions and data.

```{r, echo=FALSE, include=FALSE}
#==========================================================================
# Set up
#==========================================================================
knitr::opts_chunk$set(cache = TRUE, cache.comments = FALSE, 
                      include = TRUE, 
                      fig.width = 7, fig.height = 7,
                      fig.pos = 'H', 
                      dev = 'png', dpi = 300)

require(plyr)
require(dplyr) # For filter
require(gtools) # For permutations
require(tictoc) # For bdiag()
require(doParallel) # For post_prob_R
require(Matrix)
require(matrixStats) # for logSumExp
library(igraph) # For igraph
require(RColorBrewer)
require(knitr) # for kable()
load('../RData/TimingModel/MOD2_theta_estimates.RData')
load('../RData/TimingModel/MOD2_Posterior_samples.RData')
load('../RData/TimingModel/Combined_Time_Event.RData')

source("../Genetic_Model/Data_functions.R")
source('../Genetic_Model/iGraph_functions.R')
source("../Genetic_Model/post_prob_CLI.R") 
source("../Genetic_Model/post_prob_CLI_sequential.R") 
source("../Genetic_Model/test_Rn_compatible.R") 
source("../Genetic_Model/Data_Inflation_Functions.R")
#source('../Plotting_MS_data/PlottingFunction.R')

# The pooled MS data from BPD and VHX
load('../RData/GeneticModel/MS_data_PooledAnalysis.RData')

# Booleans that describe generation of Markdown file. 
# The large jobs should be done on the cluster

# runs the once model on those that can be directly computed
RUN_MODELS_SINGLE_SIMPLE = F           # can do on laptop (5 minutes)
# runs the model for 100 random draws from the Time model posterior on those that can be directly computed
RUN_MODELS_FULL_POSTERIOR_SIMPLE = F    # can do on laptop (2 hours)
# runs the model for 100 random draws from the Time model posterior on those that cannot be directly computed
RUN_MODELS_FULL_POSTERIOR_INFLATED = F  # cluster is better (10 hours)
# Does the false positive discovery estimation
RUN_MODELS_FALSE_POSITIVE = F           # only cluster (1 day with 42 cores)
# generates plots
CREATE_PLOTS = T

# Colour scheme
# Previous Set1 not colourblind friendly: display.brewer.all(colorblindFriendly = T)
Dark2 = brewer.pal(8, 'Dark2')

# This needs moving to time analysis: 
drug_cols3 = array(Dark2[c(4,6,1)], dim = 3, dimnames = list(c('AS','CHQ','CHQ/PMQ'))) 
drug_cols2 = array(Dark2[c(2,2,1)], dim = 3, dimnames = list(c('AS','CHQ','CHQ/PMQ')))
drug_cols_light2 = sapply(drug_cols2, adjustcolor, alpha.f = 0.5)

# Vector of states
states = c(relapse = 'L', reinfection = 'I', recrudescence = 'C')

# James, there are some odd bugs something to do with !duplicated() and the tibble...
# https://stackoverflow.com/questions/39041115/fixing-a-multiple-warning-unknown-column
# Also, Thetas_full_post has redundant column names, this is a quick solution
Combined_Time_Data <- as.data.frame(Combined_Time_Data)

## Threshold value for classification
Epsilon_upper = 0.7
Epsilon_lower = 0.3
transparent_pink_band = adjustcolor(Dark2[4], alpha.f = 0.2)

```


Summary of the data and the whole of the VHX data set versus the subset typed (in terms of number of episodes):

```{r, echo=FALSE}
writeLines(sprintf('Number of individuals with at least one episode typed: %s',
                   length(unique(MS_pooled$ID))))
writeLines(sprintf('Number of episodes typed: %s',
                   length(unique(MS_pooled$Episode_Identifier))))
writeLines(sprintf('Number of recurrences typed: %s',
                   length(unique(MS_pooled$Episode_Identifier[MS_pooled$Episode>1]))))

# Which drug arms have been typed:
writeLines('\nOverall in the dataset: breakdown by treatment group (individuals typed):')
table(MS_pooled$Treatment[!duplicated(MS_pooled$ID)])
writeLines('\nWithin VHX: breakdown by treatment group (VHX individuals typed):')
table(MS_pooled$Treatment[!duplicated(MS_pooled$ID) & 
                            1:nrow(MS_pooled)%in% grep('VHX',MS_pooled$ID)])

writeLines(sprintf('\nFrom BPD trial there are %s individuals with total of %s episodes typed (enrollment: %s; recurrent %s)',
                   length(unique(MS_pooled$ID[grep('BPD',MS_pooled$ID)])),
                   length(unique(MS_pooled$Episode_Identifier[grep('BPD',MS_pooled$ID)])), 
                   length(unique(MS_pooled[grepl('BPD',MS_pooled$ID) & MS_pooled$Episode == 1, 'Episode_Identifier'])), 
                   length(unique(MS_pooled[grepl('BPD',MS_pooled$ID) & MS_pooled$Episode > 1, 'Episode_Identifier']))))
writeLines(sprintf('From VHX trial there are %s individuals with total of %s episodes typed (enrollment: %s; recurrent %s)',
                   length(unique(MS_pooled$ID[grep('VHX',MS_pooled$ID)])),
                   length(unique(MS_pooled$Episode_Identifier[grep('VHX',MS_pooled$ID)])),
                   length(unique(MS_pooled[grepl('VHX',MS_pooled$ID) & MS_pooled$Episode == 1, 'Episode_Identifier'])), 
                   length(unique(MS_pooled[grepl('VHX',MS_pooled$ID) & MS_pooled$Episode > 1, 'Episode_Identifier']))))
```


```{r}
#=================================================================
# Were all episodes typed if a person was selected for genotyping? 
#=================================================================
MS_pooled_summary = MS_pooled[!duplicated(MS_pooled$Episode_Identifier),] # Collapse rows due to COI > 2
censored_ind = Combined_Time_Data$Censored == 1 # index censored rows in Combined_Time_Data to prevent over counting rows
uncensored_patientids = Combined_Time_Data[!censored_ind, 'patientid'] # vector of patientids excluding censored rows (inc. duplicates)
All_VHX_epi_count = table(uncensored_patientids[grepl('VHX_',uncensored_patientids)]) # Number of episodes per person VHX
All_BPD_epi_count = table(uncensored_patientids[grepl('BPD_',uncensored_patientids)]) # Number of episodes per person VHX

# No. of typed episodes per person with one or more typed episodes in VHX and BPD
no_of_typed_epi_per_person_typed_VHX = table(MS_pooled_summary$ID[grepl('VHX',MS_pooled_summary$ID)])  
no_of_typed_epi_per_person_typed_BPD = table(MS_pooled_summary$ID[grepl('BPD',MS_pooled_summary$ID)])  

# No. of total episodes per person with one or more typed episodes in VHX and BPD
no_of_epi_per_person_typed_VHX = All_VHX_epi_count[names(All_VHX_epi_count) %in% names(no_of_typed_epi_per_person_typed_VHX)]
no_of_epi_per_person_typed_BPD = All_BPD_epi_count[names(All_BPD_epi_count) %in% names(no_of_typed_epi_per_person_typed_BPD)]


length(All_VHX_epi_count)
length(All_BPD_epi_count)
#-------------------------
# VHX data set summary: breif because genotyping VHX was not exhaustive
#-------------------------
X0 = length(no_of_typed_epi_per_person_typed_VHX) # Number of people typed
ind_untyped = no_of_epi_per_person_typed_VHX != no_of_typed_epi_per_person_typed_VHX
X1 = sum(ind_untyped) # Number of people selected for genotyping but some episodes untyped
# How many untyped per person with incomplete set of episodes typed:
X2 = range(no_of_epi_per_person_typed_VHX[ind_untyped] - no_of_typed_epi_per_person_typed_VHX[ind_untyped]) 
X3 = sum(no_of_epi_per_person_typed_VHX - no_of_typed_epi_per_person_typed_VHX) # Total number untyped
writeLines(sprintf('VHX: for %s of %s VHX individual/s selected for genotyping: %s to %s of their episodes were not typed (total of %s episodes untyped summing over the %s individuals)',X1,X0,X2[1],X2[2],X3,X1))


#-------------------------
# BPD data set: comprehensive because genotyping BPD was exhaustive
#-------------------------
# How many people who experience one or more recurrences had one or more episodes genotyped? 
recurrences = All_BPD_epi_count[All_BPD_epi_count > 1]-1
indivs_who_recurred = names(recurrences)
indivs_who_were_typed = names(no_of_typed_epi_per_person_typed_BPD)
indivs_who_were_not_typed = indivs_who_recurred[!indivs_who_recurred %in% indivs_who_were_typed]

# Summary over individuals typed 
X0 = length(indivs_who_were_typed) # Number of people typed
ind_untyped = no_of_epi_per_person_typed_BPD != no_of_typed_epi_per_person_typed_BPD
X1 = sum(ind_untyped) # Number of episodes untyped
# How many untyped per person with incomplete set of episodes typed:
X2 = range(no_of_epi_per_person_typed_BPD[ind_untyped] - no_of_typed_epi_per_person_typed_BPD[ind_untyped])  
X3 = sum(no_of_epi_per_person_typed_BPD - no_of_typed_epi_per_person_typed_BPD)  # Total number untyped

# Individuals with not all episodes typed
ind_missing_typed_epi <- names(which(no_of_epi_per_person_typed_BPD != no_of_typed_epi_per_person_typed_BPD)) 

# All episodes of the BPD individuals missing one or episodes
X4 = lapply(ind_missing_typed_epi, function(x){
  ind = grepl(x, uncensored_patientids)
  Combined_Time_Data$episode[!censored_ind][ind]
})

# Typed episodes of the BPD individuals missing one or episodes
X5 = lapply(ind_missing_typed_epi, function(x){
  ind = grepl(x, MS_pooled_summary$ID)
  MS_pooled_summary$Episode[ind]
})

X6 = lapply(1:length(X5), function(i){setdiff(X4[[i]], X5[[i]])}) # Not typed episodes
X7 = sum(sapply(X6, function(x)sum(x>1))) # Not typed recurrence
```

```{r, echo=FALSE}
writeLines(paste(sprintf('BPD: of %s of the people who recurred: %s person/people with %s recurrence/s was not selected for genotyping.',
                         length(unique(indivs_who_recurred)), 
                         length(unique(indivs_who_were_not_typed)), 
                         recurrences[indivs_who_were_not_typed]), 
                 sprintf('Of %s of %s BPD individual/s selected for genotyping: %s to %s of their episodes were not typed (%s episodes summing over the %s individuals).', X1,X0,X2[1],X2[2],X3,X1), 
                 sprintf('Of the %s episodes not typed %s were recurrences.',X3, X7),
                 sprintf('In total there were %s recurrences: %s untyped.', sum(recurrences), recurrences[indivs_who_were_not_typed] + X7)))

# For plots
par(mfrow = c(1,2), family = 'serif', pty = 's') 
plot(x = as.numeric(no_of_epi_per_person_typed_VHX), y = as.numeric(no_of_typed_epi_per_person_typed_VHX), 
     xlab = 'Number of episodes per person', ylab = 'Number of typed episodes per person', main = 'VHX',
     ylim = range(no_of_epi_per_person_typed_VHX), xlim = range(no_of_epi_per_person_typed_VHX))
abline(a = 0, b = 1, lty = 'dotted')
plot(x = as.numeric(no_of_epi_per_person_typed_BPD), y = as.numeric(no_of_typed_epi_per_person_typed_BPD), 
     xlab = 'Number of episodes per person', ylab = 'Number of typed episodes per person', main = 'BPD',
     ylim = range(no_of_epi_per_person_typed_BPD), xlim = range(no_of_epi_per_person_typed_BPD))
abline(a = 0, b = 1, lty = 'dotted')
```


```{r VHX_typed_untyped}
#=================================================================
# Visual check of difference in episode counts between VHX 
# that where genetically typed and not
#=================================================================
# Condition on CQ since these were the ones selected for genotyping 
CQ_ind_epi <- Combined_Time_Data$arm_num == "CHQ" 
# vector of patientids excluding censored rows and those without CQ
uncensored_patientids = Combined_Time_Data[!censored_ind & CQ_ind_epi, 'patientid'] 
# Number of episodes per person VHX
All_VHX_epi_count = table(uncensored_patientids[grepl('VHX_',uncensored_patientids)]) 

# Condition on those that have one or more recurrence
All_VHX_rec_count = All_VHX_epi_count[All_VHX_epi_count > 1] 
x1 = no_of_epi_per_person_typed_VHX
x2 = All_VHX_rec_count[!names(All_VHX_rec_count) %in% names(x1)] # No of epi per person untyped
# setequal(names(x1), names(x2)) # Check mutually exclusive
# setequal(names(x2), unique(names(MS_pooled_summary$ID))) # Further check 
max_rec = max(All_VHX_rec_count)
```

```{r, echo=FALSE}
# Typed more heavy tailed
hist(x2, col = 'lightgray', freq = F, breaks = seq(0.5, max_rec+0.5, 1), ylim = c(0,0.35), 
     main='VHX subset', xlab = 'Number of episodes')
hist(x1, col = 'red', density = 15, add = T, freq = F, breaks = seq(0.5, max_rec+0.5, 1))
legend('topright', legend = c('Untyped', 'Typed'), fill =c('lightgray', 'red'),
       density = c(100,25), bty = 'n')
```

Summary of complexity of infection based on numbers of alleles observed. This is broken down by enrollment episodes (this is independent of drug given) and subsequent recurrences which could be drug dependent.

```{r COIs_VHX_BPD}
COIs = data.frame(t(sapply(unique(MS_pooled$Episode_Identifier), function(x){
  ind = which(MS_pooled$Episode_Identifier == x)
  c(MOI=max(MS_pooled$MOI_id[ind]),
    Enrollment = MS_pooled$Episode[ind[1]] == 1,
    Drug = MS_pooled$Treatment[ind[1]])
})))
COIs$MOI = as.numeric(COIs$MOI)
COIs$Enrollment = COIs$Enrollment=='TRUE'
COIs$PMQ = 0 
COIs$PMQ[!COIs$Enrollment & COIs$Drug=='PMQ']=1
```

```{r, echo=FALSE}
par(las=1, mfrow=c(1,2))

# Distribution of enrollment COIs
hist(COIs$MOI[COIs$Enrollment], freq = F, breaks = seq(0.5,4.5, by=1),
     xlab = 'COIs', ylab = '% of enrollment episodes',
     main='Enrollment episodes', col = 'gray', 
     yaxt='n', ylim = c(0,.75))
axis(2, at = c(0,.25,.5,.75), labels = 100*c(0,.25,.5,.75))

# Distribution of recurrent COIs by drug
hist(COIs$MOI[!COIs$Enrollment & COIs$Drug=='PMQ'], freq = F, 
     breaks = seq(0.5,4.5, by=1),
     xlab = 'COIs', ylab = '% of enrollment episodes',
     yaxt='n', ylim = c(0,.75), main='Recurrent episodes ', 
     density = 15, angle = 30, col ='blue')
axis(2, at = c(0,.25,.5,.75), labels = 100*c(0,.25,.5,.75))
hist(COIs$MOI[!COIs$Enrollment & !COIs$Drug=='PMQ'], freq = F, 
     breaks = seq(0.5,4.5, by=1),
     xlab = 'COIs', ylab = '% of enrollment episodes',add=T,
     yaxt='n', ylim = c(0,.75), main='', density = 15, 
     angle = -30, col ='grey')
axis(2, at = c(0,.25,.5,.75), labels = 100*c(0,.25,.5,.75))
legend('topright',col=c('blue','grey'),legend = c('PMQ+','No PMQ'), lwd=2)
mod = glm(MOI ~ enrollment + drug, family = 'poisson', 
          data = data.frame(MOI=COIs$MOI - 1, # need to do minus 1 so it's Poisson appropriate
                            enrollment=as.numeric(COIs$Enrollment),
                            drug = COIs$PMQ))
summary(mod)
writeLines(sprintf('Mean complexity of recurrent episodes is %s, and mean complexity of enrollment episodes is %s',
                   round(1+exp(mod$coefficients['(Intercept)']),2),
                   round(1+exp(mod$coefficients['(Intercept)']+mod$coefficients['enrollment']),2)))

# Summary of median COI per study
BPD_ind = grepl('BPD_', rownames(COIs))
VHX_ind = grepl('VHX_', rownames(COIs))
writeLines(sprintf('Median COI in VHX and BPD: %s and %s, respectively',
                   median(COIs$MOI[VHX_ind]),median(COIs$MOI[BPD_ind])))

# Summary of COIs >= 3 (important due to computational complexity)
writeLines(sprintf('%s of %s episodes (%s percent) with COI greater than or equal to 3', 
                   sum(COIs$MOI >= 3), nrow(COIs), round(100*mean(COIs$MOI >= 3))))
```


From this Poisson regression, there appears to be evidence that enrollment episodes have higher complexities of infection than recurrences. This implies that relapses are more likely to be single hypnozoite activated infections?


# Allele frequencies

First we define the set of microsatellite markers used in this analysis:
```{r}
MSs_all = c("PV.3.502","PV.3.27","PV.ms8",
            "PV.1.501","PV.ms1","PV.ms5",
            "PV.ms6","PV.ms7","PV.ms16")
```

We use a multinomial-dirichlet model with subjective weight $\omega$. $\omega = 0$ recovers the empirical allele frequencies. 

```{r, echo=F}
# Prior weight for the Dirichlet (setting weight to 0 recovers empirical freq):
D_weight_Prior = 1

# These are motif lengths: for the plotting
MSs_Motifs = list("PV.3.502"=8,'PV.3.27' = 4, 
                  "PV.ms8" = 3, "PV.1.501"= 7, 
                  "PV.ms1" = 3, "PV.ms5" = 3,
                  "PV.ms6" = 3, "PV.ms16" =3,
                  "PV.ms7" = 3)

writeLines(paste('Number of episodes used to compute frequencies:',
                 sum(MS_pooled$Episode==1 & MS_pooled$MOI_id==1)))
Ind_Primary = which(MS_pooled$Episode==1)

# Posterior Dirichlet parameter vector
Alpha_Posteriors = apply(MS_pooled[,MSs_all], 2, function(x, Ind_Primary){
  # Extract xmax 
  xmax = max(x,na.rm=T)
  # prior parameter vector (iterpolates unobserved repeat lengths < xmax)
  param_vector = array(D_weight_Prior, dim = xmax, dimnames = list(1:xmax)) 
  # observed data summarised as counts
  obs_counts = table(x[Ind_Primary]) 
  # posterior parameter vector
  param_vector[names(obs_counts)] = param_vector[names(obs_counts)] + obs_counts
  return(param_vector)
})

# Calculate posterior mean 
Fs_Combined = sapply(Alpha_Posteriors, function(x){x/sum(x)})
```

```{r}
#==================================================================
# Save a data set of monoclonal data and allele frequencies for
# relatedness estimation 
#==================================================================
monoclonal_names = rownames(COIs)[COIs$MOI == 1]
monoclonal_data = MS_pooled[MS_pooled$Episode_Identifier%in%monoclonal_names, ]
save(monoclonal_data, Fs_Combined, file = '../RData/Data_for_relatedness.RData')
```

Calculate the effective marker cardinality for each microsatellite marker using a simulation approach.

```{r}
N = 10^6
Effective_Allele_size = list()
for(ms in MSs_all){
  n_obs_alleles = length(table(MS_pooled[Ind_Primary,ms]))
  draw1 = sample(x = names(Fs_Combined[[ms]]), replace = T, size = N, prob = Fs_Combined[[ms]])
  draw2 = sample(x = names(Fs_Combined[[ms]]), replace = T, size = N, prob = Fs_Combined[[ms]])
  x = mean(draw1 == draw2)
  n = 1/x
  writeLines(sprintf('The effective cardinality for %s with %s observed alleles is %s', ms, n_obs_alleles, round(n,2)))
  Effective_Allele_size[[ms]] = round(n,1)
}
# The mean and range in our data set
writeLines(sprintf('The mean effective marker cardinality is %s, range: %s to %s', 
                   round(mean(unlist(Effective_Allele_size)),2), 
                   round(min(unlist(Effective_Allele_size)),2), 
                   round(max(unlist(Effective_Allele_size)),2)))
```



## Plotting allele frequencies

These are the mean posterior allele frequencies (dots) and 95\% credible intervals (bars) given pooled enrollment data and $\omega=$ `D_weight_Prior`.  

```{r AlleleFrequencies, echo=F}
if(CREATE_PLOTS){
  par(mfrow=c(3,3), las=1, bty='n', cex.axis=1.2, family = 'serif')
  for(ms in MSs_all){ # As bars
    K = length(Fs_Combined[[ms]]) # Cardinality of ms 
    xs = rdirichlet(n = 1000, alpha = Alpha_Posteriors[[ms]]) # Sample from posterior
    # MC approximation of 0.975 percentile expressed as percentage
    YMAX = max(apply(xs, 2, quantile, probs = .975)) # Round up to nearest decimal
    # Number of observations
    N_MS = length(unique(MS_pooled$ID[MS_pooled$Episode==1 & !is.na(MS_pooled[,ms])]))
    plot(NULL, panel.first = grid(),
         main = sprintf('%s', ms), font.main = 3, 
         pch = 18, ylim=c(0,YMAX), xlim = c(1,ncol(xs)), 
         ylab='', xlab='', yaxt='n')
    abline(h=1/K, lty = 'dashed', lwd = 1)
    title(ylab='Frequency', line = 3.25)
    title(xlab= sprintf('Repeat length (motif length = %s)', MSs_Motifs[[ms]]), line = 2.5)
    # made the number of individuals N: same as appendix notation
    mtext(side = 3, text = sprintf('(N = %s)',N_MS), line = 0.5, cex = 0.65, adj = 0)
    mtext(side = 3, text = sprintf('(n* = %s)',Effective_Allele_size[[ms]]), 
          line = 0.5, cex = 0.65, adj = 1)
    for(k in 1:ncol(xs)){
      lines(rep(k,2), quantile(xs[,k],probs = c(0.025,0.975)))
    }
    points(1:length(Fs_Combined[[ms]]),Fs_Combined[[ms]],pch=18) # Point estimates: mean 
    axis(2, round(seq(0, round(YMAX,2), length.out = 3),2))
  }
}
```


# Computing the probability of relatedness across infections

The approach is Bayesian and consists of the following:

* A prior probability vector for the recurrence state from the time-to-event model
* An allele frequency estimate from the posterior distribution of allele frequencies
* A likelihood based on the genetic data of being a *relapse*, a *recrudescence*, or a *reinfection* given the observed microsatellite data.

The following iterates through each individual and computes the probability of relatedness states.

```{r}
# # We also remove MS data for which there are no recurrent data
N_episodes_typed = table(MS_pooled$ID[!duplicated(MS_pooled$Episode_Identifier)])
MS_pooled = filter(MS_pooled, ID %in% names(N_episodes_typed[N_episodes_typed>1]))
# recreate pooled summary dataset
MS_pooled_summary = MS_pooled[!duplicated(MS_pooled$Episode_Identifier),] 

writeLines(sprintf('Number of individuals with at least two typed episodes analysed: %s',
                   length(unique(MS_pooled$ID))))
writeLines(sprintf('Number of episodes in individuals with at least two typed episodes analysed: %s',
                   length(unique(MS_pooled$Episode_Identifier))))
writeLines(sprintf('Number of typed recurrences analysed: %s',
                   length(unique(MS_pooled$Episode_Identifier[MS_pooled$Episode>1]))))
```


## Load the time-to-event priors

```{r}
inds = grepl('mean_theta', colnames(Mod2_ThetaEstimates)) # Extract mean
p = data.frame(Episode_Identifier = Mod2_ThetaEstimates$Episode_Identifier, 
               Mod2_ThetaEstimates[,inds],
               stringsAsFactors = F) # Reformat
colnames(p) = gsub(pattern = 'Recrudescence_mean_theta',replacement = 'C',x = colnames(p))
colnames(p) = gsub(pattern = 'Relapse_mean_theta',replacement = 'L',x = colnames(p))
colnames(p) = gsub(pattern = 'ReInfection_mean_theta',replacement = 'I',x = colnames(p))

genetic_AND_time_data_eps = intersect(p$Episode_Identifier, MS_pooled$Episode_Identifier)
p = p[p$Episode_Identifier %in% genetic_AND_time_data_eps,] # Only need priors for those with genetic data

# Extract posterior estimates only if running full posterior simple or inflated
if(RUN_MODELS_FULL_POSTERIOR_SIMPLE | RUN_MODELS_FULL_POSTERIOR_INFLATED){
  Post_samples_matrix = Post_samples_matrix[Post_samples_matrix$Episode_Identifier %in% genetic_AND_time_data_eps,]
}
```


## Computation using full dataset 

We use all 9MS markers (when available).

```{r, include=FALSE}
if(RUN_MODELS_SINGLE_SIMPLE){
  #===============================================
  # Run (with time-to-event)
  #===============================================
  # Approx 100 secs per full model run
  tic()
  thetas_9MS = post_prob_CLI(MSdata = MS_pooled, Fs = Fs_Combined, 
                             p = p, cores = 42, verbose = F) 
  thetas_9MS$Episode_Identifier = rownames(thetas_9MS)
  save(thetas_9MS, file = '../RData/GeneticModel/thetas_9MS.RData')
  toc()
  
  #===============================================
  # Run (without time-to-event)
  #===============================================
  tic()
  thetas_9MS_Tagnostic = post_prob_CLI(MSdata = MS_pooled, Fs = Fs_Combined, 
                                       cores = 42, verbose = F)
  thetas_9MS_Tagnostic$Episode_Identifier = rownames(thetas_9MS_Tagnostic)
  save(thetas_9MS_Tagnostic, file = '../RData/GeneticModel/thetas_9MS_Tagnostic.RData')
  toc()
} else {
  load('../RData/GeneticModel/thetas_9MS.RData')
  load('../RData/GeneticModel/thetas_9MS_Tagnostic.RData')
}
```

### Full posterior computation

```{r, include=FALSE}
if(RUN_MODELS_FULL_POSTERIOR_SIMPLE){
  #===============================================
  # Run full Bayesian with sampling at random from time prior (treats time posterior as discrete uniform prior)
  # takes about 1.1 hour on 6 cores
  # The outer loop is not parallelised - suboptimal coding
  #===============================================
  
  Ksamples = max(length(grep('C',colnames(Post_samples_matrix))), 100)
  tic()
  Thetas_full_post = foreach(ss = 1:Ksamples, .combine = cbind) %do% {
    
    # draw a random distribution over the allele frequencies from posterior
    Fs_random = lapply(Alpha_Posteriors, rdirichlet, n=1)
    # I'm not sure if this is needed - how are these called deep inside?
    for(i in 1:length(Fs_random)){
      names(Fs_random[[i]]) = 1:length(Fs_random[[i]])
    }
    # take the ss sample from the time prior
    indices = c(grep('C',colnames(Post_samples_matrix))[ss],
                grep('L',colnames(Post_samples_matrix))[ss],
                grep('I',colnames(Post_samples_matrix))[ss],
                grep('Episode_Identifier',colnames(Post_samples_matrix)))
    p = Post_samples_matrix[,indices]
    # run the model
    thetas_9MS = post_prob_CLI(MSdata = MS_pooled, Fs = Fs_random, 
                               p = p, cores = 42, verbose = F) 
    thetas_9MS$Episode_Identifier = rownames(thetas_9MS)
    thetas_9MS
  }
  
  save(Thetas_full_post, file = '../RData/GeneticModel/Full_Posterior_Model_samples.RData')
  toc()
} else {
  load(file = '../RData/GeneticModel/Full_Posterior_Model_samples.RData')
}
```


# Plot results


```{r}
# Output of time-to-event model (sorted by episode number s.t. columns correspond)
Time_Estimates_1 = filter(Mod2_ThetaEstimates, Episode_Identifier %in% thetas_9MS$Episode_Identifier)
Time_Estimates_1 = arrange(Time_Estimates_1, Episode_Identifier)

# Outputs of genetic model w/wo time prior 
# sorted by episode number s.t. columns correspond and drug added
thetas_9MS = arrange(thetas_9MS, Episode_Identifier)
thetas_9MS_Tagnostic = arrange(thetas_9MS_Tagnostic, Episode_Identifier)
thetas_9MS$drug = Time_Estimates_1$arm_num # Add drug
thetas_9MS_Tagnostic$drug = Time_Estimates_1$arm_num # Add drug


# Extract BPD only for BPD only plots
BPD_data = Thetas_full_post[grep('BPD',rownames(Thetas_full_post)),]
Thetas_BPD = thetas_9MS[grep('BPD', thetas_9MS$Episode_Identifier),]

# Extract prior used in absence of time-to-event
Time_agnostic_p = as.list(formals(post_prob_CLI)$p)
```

```{r, echo=FALSE}
# Highlight some specific examples (useful for slides when presenting work)
Hightlight_BPD_examples = F
if(Hightlight_BPD_examples){
  example_inds = grepl('BPD_644_', Thetas_BPD$Episode_Identifier) | 
    grepl('BPD_598_', Thetas_BPD$Episode_Identifier) | 
    grepl('BPD_562_', Thetas_BPD$Episode_Identifier) |
    grepl('BPD_53_', Thetas_BPD$Episode_Identifier) 
  example_ids = Thetas_BPD$Episode_Identifier[example_inds]
  example_inds_times = MS_pooled_summary$timeSinceLastEpisode[MS_pooled_summary$Episode_Identifier %in% example_ids]
  Tagnostic_example_inds = thetas_9MS_Tagnostic$Episode_Identifier %in% example_ids
  thetas9MS_example_inds = thetas_9MS$Episode_Identifier %in% example_ids
  Time1_example_inds = Time_Estimates_1$Episode_Identifier %in% example_ids
}
```

## Going from time-to-event prior to posterior

Plotted by radical cure versus no radical cure, as that is the most informative distinction here.

```{r Supplementary_TimeEffect_onPosterior, fig.width=7, fig.height=7, echo=FALSE}
if(CREATE_PLOTS){
  par(mfrow=c(2,2),las=1, bty='n',family='serif')
  
  
  # Time agnostic versus full posterior 
  plot(x = thetas_9MS_Tagnostic$L, 
       y = thetas_9MS$L, 
       log = 'xy', 
       xlab = 'Genetic given time-agnostic prior', 
       ylab = 'Genetic given time-to-event based prior',
       panel.first = grid(),
       col = drug_cols2[thetas_9MS$drug], 
       main = 'Probability of relapse',pch=20)
  abline(a = 0, b = 1)
  
  polygon(x = c(range(thetas_9MS_Tagnostic$L, na.rm = T),
                rev(range(thetas_9MS_Tagnostic$L, na.rm = T))),
          y = c(Epsilon_lower,
                Epsilon_lower,
                Epsilon_upper,
                Epsilon_upper),
          col = transparent_pink_band, border = NA)
  
  polygon(y = c(range(thetas_9MS$L, na.rm = T),
                rev(range(thetas_9MS$L, na.rm = T))),
          x = c(Epsilon_lower,
                Epsilon_lower,
                Epsilon_upper,
                Epsilon_upper),
          col = transparent_pink_band, border = NA)
  
  
  # Annotate by examples
  if(Hightlight_BPD_examples){
    points(x = log10(thetas_9MS_Tagnostic$L[Tagnostic_example_inds]), 
           y = log10(thetas_9MS$L[thetas9MS_example_inds]), 
           pch=1, cex = 1.5, col='black')
    text(x = log10(thetas_9MS_Tagnostic$L[Tagnostic_example_inds]), 
         y = log10(thetas_9MS$L[thetas9MS_example_inds]), 
         labels = example_ids, pos = 2, cex = 0.7)
  }
  
  plot(x = thetas_9MS_Tagnostic$I, 
       y = thetas_9MS$I, 
       xlab = 'Genetic given time-agnostic prior', 
       ylab = 'Genetic given time-to-event based prior',
       log = 'xy', 
       panel.first = grid(),
       col = drug_cols2[thetas_9MS$drug], 
       main = 'Probability of reinfection',pch=20)
  abline(a = 0, b = 1)
  legend('topleft', legend = c('No PMQ', 'PMQ+'), inset = 0.05, 
         col = drug_cols2[2:3], pch = 20, bty = 'o', bg = 'white')
  
  # polygon(x = c(range(thetas_9MS_Tagnostic$I, na.rm = T),
  #               rev(range(thetas_9MS_Tagnostic$I, na.rm = T))),
  #         y = c(Epsilon_lower,
  #               Epsilon_lower,
  #               Epsilon_upper,
  #               Epsilon_upper),
  #         col = transparent_pink_band, border = NA)
  # 
  # polygon(y = c(range(thetas_9MS$I, na.rm = T),
  #               rev(range(thetas_9MS$I, na.rm = T))),
  #         x = c(Epsilon_lower,
  #               Epsilon_lower,
  #               Epsilon_upper,
  #               Epsilon_upper),
  #         col = transparent_pink_band, border = NA)
  
  # Prior versus full posterior
  plot(x = Time_Estimates_1$Relapse_mean_theta,
       y = thetas_9MS$L,
       xlab = 'Time-to-event (genetic time-to-event based prior)', 
       ylab = 'Genetic given time-to-event-based prior',
       log = 'xy', 
       panel.first = grid(),
       col= drug_cols2[thetas_9MS$drug],
       main = 'Probability of relapse',pch=20)
  abline(a = 0, b = 1)
  
  polygon(x = c(range(Time_Estimates_1$Relapse_mean_theta, na.rm = T),
                rev(range(Time_Estimates_1$Relapse_mean_theta, na.rm = T))),
          y = c(Epsilon_lower,
                Epsilon_lower,
                Epsilon_upper,
                Epsilon_upper),
          col = transparent_pink_band, border = NA)

  polygon(y = c(range(thetas_9MS$L, na.rm = T),
                rev(range(thetas_9MS$L, na.rm = T))),
          x = c(Epsilon_lower,
                Epsilon_lower,
                Epsilon_upper,
                Epsilon_upper),
          col = transparent_pink_band, border = NA)
  
  # Annotate by examples
  if(Hightlight_BPD_examples){
    points(x = log10(Time_Estimates_1$Relapse_mean_theta[Time1_example_inds]), 
           y = log10(thetas_9MS$L[thetas9MS_example_inds]), 
           pch=1, cex = 1.5, col='black')
    text(x = log10(Time_Estimates_1$Relapse_mean_theta[Time1_example_inds]), 
         y = log10(thetas_9MS$L[thetas9MS_example_inds]), 
         labels = example_ids, pos = 2, cex = 0.7)
  }
  plot(x = Time_Estimates_1$ReInfection_mean_theta,
       y = thetas_9MS$I,
       log = 'xy', 
       xlab = 'Time-to-event (genetic time-to-event based prior)', 
       ylab = 'Genetic given time-to-event-based prior',
       panel.first = grid(), 
       col= drug_cols2[thetas_9MS$drug], 
       main = 'Probability of reinfection', pch=20)
  abline(a = 0, b = 1)
  
  # polygon(x = c(range(Time_Estimates_1$ReInfection_mean_theta, na.rm = T),
  #               rev(range(Time_Estimates_1$ReInfection_mean_theta, na.rm = T))),
  #         y = c(Epsilon_lower,
  #               Epsilon_lower,
  #               Epsilon_upper,
  #               Epsilon_upper),
  #         col = transparent_pink_band, border = NA)
  # 
  # polygon(y = c(range(thetas_9MS$I, na.rm = T),
  #               rev(range(thetas_9MS$I, na.rm = T))),
  #         x = c(Epsilon_lower,
  #               Epsilon_lower,
  #               Epsilon_upper,
  #               Epsilon_upper),
  #         col = transparent_pink_band, border = NA)
  
  
}
```

Probability of states, ordered from most to least likely:

```{r, fig.height=12, echo=FALSE}
if(CREATE_PLOTS){
  par(mfrow=c(3,2),las=1, bty='n',family = 'serif')
  for(state in states){
    
    # Get ranks 
    ordered = sort.int(thetas_9MS[,state], decreasing = TRUE, index.return = TRUE)
    ordered_Tagn = sort.int(thetas_9MS_Tagnostic[,state], decreasing = TRUE, index.return = TRUE)
    
    # Plot genetic with time-to-event
    plot(ordered$x, 
         panel.first = grid(), 
         pch=20, 
         col = drug_cols2[thetas_9MS$drug[ordered$ix]],
         xlab = 'Rank', 
         ylab = sprintf('Probability of %s state', state),
         main = 'Genetic model given time-to-event prior')
    legend('topright',col = drug_cols2[2:3], bty = 'n', legend = c('No PMQ','PMQ+'),pch=20)
    
    # Add CIs
    ind = grepl(state,colnames(Thetas_full_post)) & !grepl('Episode_Identifier',colnames(Thetas_full_post)) 
    X = Thetas_full_post[ordered$i,ind]
    CI = t(apply(X, 1, quantile, probs = c(0.025, 0.975), na.rm=T))
    
    # I don't think the bars at the end of CIs are necessary
    for(i in 1:length(ordered$x)){
      if(!any(is.na(CI[i,])) & diff(CI[i,]) > 0.005) segments(x0 = i, y0 = CI[i,1], x1 = i, y1 = CI[i,2], 
                                                              col = drug_cols2[thetas_9MS$drug[ordered$ix[i]]])
    }
    
    
    # Plot Genetic with time-agnostic prior
    plot(ordered_Tagn$x, 
         panel.first = grid(), 
         pch=20, cex=.8,
         col = drug_cols2[thetas_9MS$drug[ordered$ix]],
         xlab = 'Rank', 
         ylab = sprintf('Probability of %s state',state),
         main = 'Genetic model given time-agnostic prior')
    legend('topright',col = drug_cols2[2:3], bty = 'n', legend = c('No PMQ','PMQ+'),pch=20)
  }
}
```


# BPD Final plot


```{r BPD_efficacy, fig.width=10,echo=FALSE}
if(CREATE_PLOTS){
  par(mfrow=c(1,2),las=1, bty='n', family = 'serif')
  ordered = sort.int(Thetas_BPD$L, decreasing = TRUE, index.return = TRUE)
  
  plot(ordered$x, 
       panel.first = grid(), 
       pch=20, col = drug_cols2[Thetas_BPD$drug[ordered$ix]],
       xlab = 'Rank', ylab = 'Probability of relapse',
       main = '')
  
  X = BPD_data[ordered$ix,grep('L',colnames(BPD_data)),]
  CI = t(cbind(apply(X, 1, quantile, probs = c(0.025, 0.975))))
  
  for(i in 1:length(ordered$x)){
    if(diff(CI[i,]) > 0.005) segments(i,CI[i,1],i,CI[i,2], 
                                      col=drug_cols2[Thetas_BPD$drug[ordered$ix[i]]])
  }
  
  writeLines(sprintf('The mean percentage of recurrences which are estimated to be relapses is %s%%',
                     round(100*sum(Thetas_BPD$L + Thetas_BPD$C)/nrow(Thetas_BPD))))
  
  
  plot(NULL,
       xlim=c(0,max(MS_pooled_summary$timeSinceLastEpisode,na.rm=T)), 
       ylim=c(0,1),
       ylab='Probability of relapse', 
       xlab = 'Time since last episode')
  
  for(i in 1:length(ordered$x)){
    kk = ordered$ix[i]
    x_time = MS_pooled_summary$timeSinceLastEpisode[Thetas_BPD$Episode_Identifier[kk]==
                                                      MS_pooled_summary$Episode_Identifier]
    points(x_time,
           Thetas_BPD$L[kk], pch=20, col = drug_cols2[Thetas_BPD$drug[ordered$ix[i]]])
    
    if(diff(CI[i,]) > 0.005) {
      segments(x_time,CI[i,1],x_time,CI[i,2], 
               #length = 0.02,angle = 90, code = 3,
               col=drug_cols_light2[Thetas_BPD$drug[ordered$ix[i]]])
    }
  }
  
  if(Hightlight_BPD_examples){
    points(x = example_inds_times,
           y = Thetas_BPD$L[example_inds], pch=1, cex = 1.5, col='black')
    text(x = example_inds_times, 
         y = Thetas_BPD$L[example_inds],
         labels = example_ids, pos =3)
  }
}
```


# Extra computations for VHX: too complex episodes


We remove the IDs that can be straightforwardly calculated:

```{r}
ind_calculated = which(MS_pooled_summary$Episode_Identifier %in% thetas_9MS$Episode_Identifier)
IDs_calculated = unique(MS_pooled_summary$ID[ind_calculated])
IDs_remaining = unique(MS_pooled_summary$ID[! MS_pooled_summary$ID %in% IDs_calculated])
```

We blow up the pooled analysis into all pairs within individuals:

```{r, include=FALSE}
MS_inflate = reformat_MSdata(filter(MS_pooled, ID %in% IDs_remaining), MSs = MSs_all)
MS_inflated = Inflate_into_pairs(MS_data = MS_inflate)
```


```{r, include=FALSE}
if(RUN_MODELS_FULL_POSTERIOR_INFLATED){  
  
  all_rec_eps = unique(MS_inflated$Episode_Identifier[MS_inflated$Episode==2])
  P_matrix = data.frame(array(dim = c(length(all_rec_eps),4)))
  colnames(P_matrix) = c('Episode_Identifier','C','I','L')
  P_matrix$Episode_Identifier = all_rec_eps
  Ksamples = min(length(grep('C',colnames(Post_samples_matrix))), 20)
  
  K_results = sum(!duplicated(MS_inflated$Episode_Identifier[MS_inflated$Episode>1]))
  
  tic()
  Res_total=foreach(ss = 1:Ksamples, .combine = cbind) %do% {
    
    # draw a random distribution over the allele frequencies from posterior
    Fs_random = lapply(Alpha_Posteriors, rdirichlet, n=1)
    # I'm not sure if this is needed - how are these called deep inside?
    for(i in 1:length(Fs_random)){
      names(Fs_random[[i]]) = 1:length(Fs_random[[i]])
    }
    # take the ss sample from the time prior
    rec_states = c('C','L','I')
    indices = c(grep(rec_states[1],colnames(Post_samples_matrix))[ss],
                grep(rec_states[2],colnames(Post_samples_matrix))[ss],
                grep(rec_states[3],colnames(Post_samples_matrix))[ss])
    for(ep in all_rec_eps){
      i = which(P_matrix$Episode_Identifier==ep)
      j = which(MS_inflated$Episode_Identifier==ep)[1]
      k = which(Post_samples_matrix$Episode_Identifier == paste(MS_inflated$ID_True[j],
                                                                MS_inflated$Second_EpNumber[j],
                                                                sep='_'))
      P_matrix[i,rec_states] = Post_samples_matrix[k,indices]
    }
    # This is about 6 minutes per run
    Res = post_prob_CLI(MSdata = MS_inflated, 
                        Fs = Fs_random, 
                        p = P_matrix,
                        UpperComplexity = 10^6, 
                        verbose = F,
                        cores = 42)
    Res
  }
  save(Res_total, file = 'FullPosterior_INF.bigRData') # save for safety
  toc()
  # Summarise the results, all CIs are 80%
  # Save this summary
  Results_Inflated = data.frame(Episode_Identifier = rownames(Res_total),
                                C_mean = apply(Res_total[,grep('C',colnames(Res_total))],1,
                                               quantile,probs = 0.5,na.rm=T),
                                C_min = apply(Res_total[,grep('C',colnames(Res_total))],1,
                                              quantile,probs = 0.1,na.rm=T),
                                C_max = apply(Res_total[,grep('C',colnames(Res_total))],1,
                                              quantile,probs = 0.9,na.rm=T),
                                L_min = apply(Res_total[,grep('L',colnames(Res_total))],1,
                                              quantile,probs = 0.1,na.rm=T),
                                L_max = apply(Res_total[,grep('L',colnames(Res_total))],1,
                                              quantile,probs = 0.9,na.rm=T),
                                L_mean = apply(Res_total[,grep('L',colnames(Res_total))],1,
                                               quantile,probs = 0.5,na.rm=T),
                                I_mean = apply(Res_total[,grep('I',colnames(Res_total))],1,
                                               quantile,probs = 0.5,na.rm=T),
                                I_min = apply(Res_total[,grep('I',colnames(Res_total))],1,
                                              quantile,probs = 0.1,na.rm=T),
                                I_max = apply(Res_total[,grep('I',colnames(Res_total))],1,
                                              quantile,probs = 0.9,na.rm=T))
  save(Results_Inflated, file = '../RData/GeneticModel/Full_Posterior_Inflated.RData')
} else {
  #load('FullPosterior_INF.bigRData')
  load('../RData/GeneticModel/Full_Posterior_Inflated.RData')
}
```


Construct adjacency graphs and compute probabilities of relapse and reinfection.

```{r}
MS_pooled_summary$L_or_C_state = MS_pooled_summary$TotalEpisodes = NA
MS_pooled_summary$L_lower = MS_pooled_summary$L_upper = MS_pooled_summary$L_mean = NA
MS_pooled_summary$C_lower = MS_pooled_summary$C_upper = MS_pooled_summary$C_mean = NA
MS_pooled_summary$I_lower = MS_pooled_summary$I_upper = MS_pooled_summary$I_mean = NA
# Arrange by complexity
# Get single rows per episode (throw away the extra MOI information)
MS_inflated_summary = MS_inflated[!duplicated(MS_inflated$Episode_Identifier) & 
                                    MS_inflated$Episode==2,]
Results_Inflated$Episode_Identifier = as.character(Results_Inflated$Episode_Identifier)
for(i in 1:nrow(MS_inflated_summary)){
  if(!length(which(MS_inflated_summary$Episode_Identifier[i] == 
                   Results_Inflated$Episode_Identifier))>0){
    MS_inflated_summary = MS_inflated_summary[-i,]
    print('removing')
  }
}
Results_Inflated$ID_True = MS_inflated_summary$ID_True
Results_Inflated$First_EpNumber = MS_inflated_summary$First_EpNumber
Results_Inflated$Second_EpNumber = MS_inflated_summary$Second_EpNumber


# Iterate through the ones we can calculate in one go
episodes_full_model = unique(Thetas_full_post$Episode_Identifier)
cols_remove = grep('Episode_Identifier', colnames(Thetas_full_post))
Thetas_full_post = Thetas_full_post[, -cols_remove]
for(ep in episodes_full_model){
  ind1 = (MS_pooled_summary$Episode_Identifier==ep)
  ind2 = rownames(Thetas_full_post)==ep
  
  ## Summaries for relapse
  MS_pooled_summary$L_upper[ind1] = quantile(unlist(Thetas_full_post[ind2, grep('L',colnames(Thetas_full_post))]),
                                             probs=0.9, na.rm = T)
  MS_pooled_summary$L_lower[ind1] = quantile(unlist(Thetas_full_post[ind2,
                                                                     grep('L',colnames(Thetas_full_post))]),
                                             probs=0.1, na.rm = T)
  MS_pooled_summary$L_mean[ind1] = quantile(unlist(Thetas_full_post[ind2, grep('L',colnames(Thetas_full_post))]),
                                            probs=0.5, na.rm = T)
  
  ## Summaries for recrudescence
  MS_pooled_summary$C_upper[ind1] = quantile(unlist(Thetas_full_post[ind2, grep('C',colnames(Thetas_full_post))]),
                                             probs=0.9, na.rm = T)
  MS_pooled_summary$C_lower[ind1] = quantile(unlist(Thetas_full_post[ind2, grep('C',colnames(Thetas_full_post))]),
                                             probs=0.1, na.rm = T)
  MS_pooled_summary$C_mean[ind1] = quantile(unlist(Thetas_full_post[ind2, grep('C',colnames(Thetas_full_post))]),
                                            probs=0.5, na.rm = T)
  
  ## Summaries for reinfection
  MS_pooled_summary$I_upper[ind1] = quantile(unlist(Thetas_full_post[ind2, grep('I',colnames(Thetas_full_post))]),
                                             probs=0.9, na.rm = T)
  MS_pooled_summary$I_lower[ind1] = quantile(unlist(Thetas_full_post[ind2, grep('I',colnames(Thetas_full_post))]),
                                             probs=0.1, na.rm = T)
  MS_pooled_summary$I_mean[ind1] = quantile(unlist(Thetas_full_post[ind2, grep('I',colnames(Thetas_full_post))]),
                                            probs=0.5, na.rm = T)
  
  # Just going to classify on relapse versus reinfection
  if(!is.na(MS_pooled_summary$L_upper[ind1])){
    if(MS_pooled_summary$L_upper[ind1] < Epsilon_lower){
      MS_pooled_summary$L_or_C_state[ind1] = 'I'
    } else if(MS_pooled_summary$L_lower[ind1] > Epsilon_upper){
      MS_pooled_summary$L_or_C_state[ind1] = 'L'
    } else {
      MS_pooled_summary$L_or_C_state[ind1] = 'Uncertain'
    }
  } else {
    MS_pooled_summary$L_or_C_state[ind1] = NA
  }
}
####***************####################
# Now iterate through the complex ones
for(i in 1:length(IDs_remaining)){
  id = IDs_remaining[i]
  Doubles_Thetas = filter(Results_Inflated, ID_True==id)
  
  for(ep in unique(Doubles_Thetas$Second_EpNumber)){
    ind1 = which(MS_pooled_summary$ID==id & MS_pooled_summary$Episode==ep)
    ind2 = which(Doubles_Thetas$Second_EpNumber == ep)
    
    MS_pooled_summary$L_lower[ind1] = mean(Doubles_Thetas$L_min[ind2],na.rm=T)
    MS_pooled_summary$L_upper[ind1] = mean(Doubles_Thetas$L_max[ind2],na.rm=T)
    MS_pooled_summary$L_mean[ind1] = mean(Doubles_Thetas$L_mean[ind2],na.rm=T)
    
    MS_pooled_summary$C_lower[ind1] = mean(Doubles_Thetas$C_min[ind2],na.rm=T)
    MS_pooled_summary$C_upper[ind1] = mean(Doubles_Thetas$C_max[ind2],na.rm=T)
    MS_pooled_summary$C_mean[ind1] = mean(Doubles_Thetas$C_mean[ind2],na.rm=T)
    
    MS_pooled_summary$I_lower[ind1] = mean(Doubles_Thetas$I_min[ind2],na.rm=T)
    MS_pooled_summary$I_upper[ind1] = mean(Doubles_Thetas$I_max[ind2],na.rm=T)
    MS_pooled_summary$I_mean[ind1] = mean(Doubles_Thetas$I_mean[ind2],na.rm=T)
    
    if(!is.na(MS_pooled_summary$L_upper[ind1])){
      if(MS_pooled_summary$L_upper[ind1] < MS_pooled_summary$L_lower[ind1]){
        writeLines(sprintf('Problem with ID %s',id))
        stop()
      }
      if(MS_pooled_summary$L_upper[ind1] < Epsilon_lower){
        MS_pooled_summary$L_or_C_state[ind1] = 'I'
      } else if(MS_pooled_summary$L_lower[ind1] > Epsilon_upper){
        MS_pooled_summary$L_or_C_state[ind1] = 'L'
      } else {
        MS_pooled_summary$L_or_C_state[ind1] = 'Uncertain'
      }
    }
  }
}

MS_pooled_summary$Drug = MS_pooled_summary$FU = NA
for(id in MS_pooled_summary$ID){
  ind = MS_pooled_summary$ID==id
  MS_pooled_summary$TotalEpisodes[ind] = max(MS_pooled_summary$Episode[ind])
  MS_pooled_summary$Drug[ind] = as.numeric(
    Combined_Time_Data$arm_num[Combined_Time_Data$patientid==id][1] == 'CHQ/PMQ') + 2
  MS_pooled_summary$FU[ind] = Combined_Time_Data$FU_time[Combined_Time_Data$patientid==id][1]
}

MS_pooled_summary$Plotting_pch_Values = 
  as.numeric(mapvalues(MS_pooled_summary$L_or_C_state, from = c('L','Uncertain','I'), to = c(17,15,1)))
MS_pooled_summary$Plotting_col_Values = 
  as.numeric(mapvalues(MS_pooled_summary$L_or_C_state, from = c('L','Uncertain','I'), to = 1:3))
```

```{r CoatneyStylePLot, echo=FALSE}
# Only the recurrences for which we can compute estimates of recurrence state
MS_final = filter(MS_pooled_summary, !is.na(L_mean))
MS_final = arrange(MS_final, desc(Drug), desc(FU), desc(TotalEpisodes))
ids = unique(MS_final$ID)

if(CREATE_PLOTS){
  ## Time series data colored by genetic STATE: classification
  mycols_states_bg = c('#FFFFFF',Dark2[4],'#000000') # colors for states - need uncertain ones as well
  mycols_states_fg = c(Dark2[8],Dark2[4],'#000000') # colors for states - need uncertain ones as well
  mycols_drugs = brewer.pal(n=3, name = 'Set1')
  par(las=1, bty='n', cex.axis=.3, mar=c(4,0,1,1), family = 'serif')
  plot(NULL, ylab = '', xlab = '',
       xlim = c(0,370), 
       ylim = c(1,length(ids)),
       xaxt='n', yaxt='n')
  mtext(text = 'Months from start of study', side = 1, line=2.5, cex=1)
  axis(1, at = seq(0,370, by=60), labels = seq(0,370, by=60)/30, cex.axis=1)
  
  for(i in 1:length(ids)){
    
    id = ids[i]
    ind = which(MS_final$ID==id)
    
    # Add the follow up time line
    lines(c(0,MS_final$FU[ind[1]]), 
          c(i,i), lty=1, 
          lwd=.5, col=drug_cols2[MS_final$Drug[ind[1]]])
    
    # For highlighting long-latency
    # Add the follow up time line
    if(id == 'VHX_235' | id == 'BPD_27'){
      lines(c(0,MS_final$FU[ind[1]]), 
            c(i,i), lty = 'dashed', 
            lwd=.5, col= 'black')
      
      cols = mycols_states_fg[MS_final$Plotting_col_Values[ind]]
      points(MS_final$timeSinceEnrolment[ind], 
             rep(i,length(ind)), 
             pch=1,cex = 1.5, 
             col='black')
    }
    
    # Add points after highlight lines
    points(MS_final$timeSinceEnrolment[ind], rep(i,length(ind)), 
           pch=MS_final$Plotting_pch_Values[ind], 
           bg=mycols_states_bg[MS_final$Plotting_col_Values[ind]],
           col=mycols_states_fg[MS_final$Plotting_col_Values[ind]], 
           cex=.6)
    
    
    
  }
  lines(x = c(0,0), y = c(0,length(ids)),lwd=3)
  
  legend('topright', col = c(mycols_states_fg[c(3,2,1)]), 
         pch = c(17,15,1), cex = 1, bty = 'n', y.intersp = 0.75, title = 'Classification',
         legend = c('Relapse','Uncertain','Reinfection'))
  
  legend('top', col = drug_cols2[2:3], 
         cex = 1, bty = 'n', y.intersp = 0.75, 
         lty = 1, lwd = 1,
         legend = c('No PMQ', 'PMQ+'))
}

writeLines(sprintf('The Coatney style plot is showing %s recurrences in %s individuals',
                   nrow(MS_final), length(ids)))
```

```{r CompleteDataPlot, fig.height=7, fig.width=7, echo=FALSE}
if(CREATE_PLOTS){
  
  par(las=1, mfcol = c(2,1), mar=c(4,4,1,1), bty='n', family = 'serif')
  MS_final = arrange(MS_final, L_mean)
  
  plot(NULL,
       xlim = c(1,nrow(MS_final)), 
       ylim = c(0,1),
       panel.first = grid(),  
       yaxt='n', ylab='', xlab = '')
  mtext(text = 'Recurrence rank',side = 1, line = 2)
  axis(2, at = c(0,.5,1))
  title(ylab = 'Relapse posterior probability', line = 3)
  polygon(x = c(-100,nrow(MS_final)+100,nrow(MS_final)+100,-100),
          y = c(Epsilon_lower,Epsilon_lower,Epsilon_upper,Epsilon_upper),
          col = transparent_pink_band, border = NA)
  
  
  segments(x0 = 1:nrow(MS_final), 
           x1 = 1:nrow(MS_final), 
           y0 = MS_final$L_lower,
           y1 = MS_final$L_upper,
           col = drug_cols_light2[MS_final$Drug])
  
  # Superimposed points
  points(1:nrow(MS_final),
         MS_final$L_mean, 
         col = drug_cols2[MS_final$Drug], 
         pch=20, cex=.8)
  
  legend(x=0.8*nrow(MS_final), y = 0.8, col = drug_cols2[2:3], 
         pch = 20, cex = 1.2, bty = 'o', border = 'white',
         legend = c('No PMQ','PMQ+'))
  
  # Against time
  plot(NULL, 
       xlim = range(MS_final$timeSinceLastEpisode), 
       ylim = c(0,1),
       panel.first = grid(), 
       xlab = '',ylab = '',yaxt='n', xaxt='n')
  axis(2, at = c(0,.5,1))
  axis(1, at = seq(0, 360, by =60), labels = seq(0, 360, by=60)/30)
  title(xlab = '', ylab = 'Relapse posterior probability', line = 3)
  mtext(text = 'Months since last episode', side = 1, line = 2)
  polygon(x = c(-10,max(MS_final$timeSinceLastEpisode)+100,
                max(MS_final$timeSinceLastEpisode)+100,-10),
          y = c(Epsilon_lower,Epsilon_lower,Epsilon_upper,Epsilon_upper),
          col = transparent_pink_band, border = NA)
  for(i in 1:nrow(MS_final)){
    
    if(abs(MS_final$L_upper[i] - MS_final$L_lower[i]) > 0.005){
      segments(x0 = MS_final$timeSinceLastEpisode[i],
               y0 = MS_final$L_lower[i],
               x1=MS_final$timeSinceLastEpisode[i],
               y1 = MS_final$L_upper[i],
               col=drug_cols_light2[MS_final$Drug[i]])
    }
    i=i+1
  }
  
  # Add points
  points(MS_final$timeSinceLastEpisode, MS_final$L_mean,
         col = drug_cols2[MS_final$Drug], 
         pch=20, cex=.8)
  
  
  legend(x=0.8*max(MS_final$timeSinceLastEpisode), y = 0.8, col = drug_cols2[2:3], 
         pch = 20, cex = 1.2, bty = 'o', border = 'white', 
         legend = c('No PMQ','PMQ+'))
  
  
}
```

Individuals who appear to relapse very late (more than 300 days after last episode):
```{r}
MS_pooled = reformat_MSdata(MS_pooled)
IDs_late_relapse = MS_final[which(MS_final$timeSinceLastEpisode>300 & MS_final$L_lower>.9),'ID']

writeLines(sprintf('The episode ids of interest are: %s',
                   MS_final[which(MS_final$timeSinceLastEpisode>300 & MS_final$L_lower>.9),
                            'Episode_Identifier']))
print(MS_pooled[MS_pooled$ID%in%IDs_late_relapse,])
```


The summaries of the final dataset. Results for all those genotyped who did not receive primaquine (artesunate or chloroquine monotherapy):
```{r, echo=FALSE}
table(MS_final$Treatment[!duplicated(MS_final$ID)]) 

ind_noPMQ = MS_final$Treatment=='AS' | MS_final$Treatment=='CHQ'
writeLines(sprintf('In no-primaquine individuals, the weighted average of relapse is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$L_mean[ind_noPMQ])/sum(ind_noPMQ),1),
                   round(100*sum(MS_final$L_lower[ind_noPMQ])/sum(ind_noPMQ),1),
                   round(100*sum(MS_final$L_upper[ind_noPMQ])/sum(ind_noPMQ),1),
                   sum(ind_noPMQ)))

writeLines(sprintf('In no-primaquine individuals, the weighted average of recrudescences is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$C_mean[ind_noPMQ],na.rm=T)/sum(ind_noPMQ),1),
                   round(100*sum(MS_final$C_lower[ind_noPMQ],na.rm=T)/sum(ind_noPMQ),1),
                   round(100*sum(MS_final$C_upper[ind_noPMQ],na.rm=T)/sum(ind_noPMQ),1),
                   sum(ind_noPMQ)))

writeLines(sprintf('In no-primaquine individuals, the weighted average of reinfections is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$I_mean[ind_noPMQ],na.rm=T)/sum(ind_noPMQ),1),
                   round(100*sum(MS_final$I_lower[ind_noPMQ],na.rm=T)/sum(ind_noPMQ),1),
                   round(100*sum(MS_final$I_upper[ind_noPMQ],na.rm=T)/sum(ind_noPMQ),1),
                   sum(ind_noPMQ)))

```

Results for all those genotyped who did receive primaquine (VHX and BPD studies combined):
```{r, echo=FALSE}
ind_PMQ = MS_final$Treatment=='PMQ'
writeLines(sprintf('In primaquine treated individuals, the weighted average of relapses is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$L_mean[ind_PMQ])/sum(ind_PMQ),1),
                   round(100*sum(MS_final$L_lower[ind_PMQ])/sum(ind_PMQ),1),
                   round(100*sum(MS_final$L_upper[ind_PMQ])/sum(ind_PMQ),1),
                   sum(ind_PMQ)))

writeLines(sprintf('In primaquine treated individuals, the weighted average of recrudescences is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$C_mean[ind_PMQ],na.rm=T)/sum(ind_PMQ),1),
                   round(100*sum(MS_final$C_lower[ind_PMQ],na.rm=T)/sum(ind_PMQ),1),
                   round(100*sum(MS_final$C_upper[ind_PMQ],na.rm=T)/sum(ind_PMQ),1),
                   sum(ind_PMQ)))

writeLines(sprintf('In primaquine treated individuals, the weighted average of reinfections is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$I_mean[ind_PMQ],na.rm=T)/sum(ind_PMQ),1),
                   round(100*sum(MS_final$I_lower[ind_PMQ],na.rm=T)/sum(ind_PMQ),1),
                   round(100*sum(MS_final$I_upper[ind_PMQ],na.rm=T)/sum(ind_PMQ),1),
                   sum(ind_PMQ)))
```

Results for all those genotyped who did receive primaquine, only in the VHX study (unknown denominator)
```{r, echo=FALSE}
ind_PMQ_VHX = intersect(which(MS_final$Treatment=='PMQ'), grep('VHX',MS_final$ID))
writeLines(sprintf('In primaquine treated individuals (VHX), the weighted average of relapses is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$L_mean[ind_PMQ_VHX])/length(ind_PMQ_VHX),1),
                   round(100*sum(MS_final$L_lower[ind_PMQ_VHX])/length(ind_PMQ_VHX),1),
                   round(100*sum(MS_final$L_upper[ind_PMQ_VHX])/length(ind_PMQ_VHX),1),
                   length(ind_PMQ_VHX)))

writeLines(sprintf('In primaquine treated individuals (VHX), the weighted average of recrudescences is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$C_mean[ind_PMQ_VHX],na.rm=T)/length(ind_PMQ_VHX),1),
                   round(100*sum(MS_final$C_lower[ind_PMQ_VHX],na.rm=T)/length(ind_PMQ_VHX),1),
                   round(100*sum(MS_final$C_upper[ind_PMQ_VHX],na.rm=T)/length(ind_PMQ_VHX),1),
                   length(ind_PMQ_VHX)))

writeLines(sprintf('In primaquine treated individuals (VHX), the weighted average of reinfections is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$I_mean[ind_PMQ_VHX],na.rm=T)/length(ind_PMQ_VHX),1),
                   round(100*sum(MS_final$I_lower[ind_PMQ_VHX],na.rm=T)/length(ind_PMQ_VHX),1),
                   round(100*sum(MS_final$I_upper[ind_PMQ_VHX],na.rm=T)/length(ind_PMQ_VHX),1),
                   length(ind_PMQ_VHX)))

```

Results for all those genotyped who did receive primaquine, only in the BPD study (known denominator)
```{r, echo=FALSE}
ind_PMQ_BPD = intersect(which(MS_final$Treatment=='PMQ'), grep('BPD',MS_final$ID))
writeLines(sprintf('In primaquine treated individuals (BPD), the weighted average of relapses is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$L_mean[ind_PMQ_BPD])/length(ind_PMQ_BPD),1),
                   round(100*sum(MS_final$L_lower[ind_PMQ_BPD])/length(ind_PMQ_BPD),1),
                   round(100*sum(MS_final$L_upper[ind_PMQ_BPD])/length(ind_PMQ_BPD),1),
                   length(ind_PMQ_BPD)))

writeLines(sprintf('In primaquine treated individuals (BPD), the weighted average of recrudescences is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$C_mean[ind_PMQ_BPD],na.rm=T)/length(ind_PMQ_BPD),1),
                   round(100*sum(MS_final$C_lower[ind_PMQ_BPD],na.rm=T)/length(ind_PMQ_BPD),1),
                   round(100*sum(MS_final$C_upper[ind_PMQ_BPD],na.rm=T)/length(ind_PMQ_BPD),1),
                   length(ind_PMQ_BPD)))

writeLines(sprintf('In primaquine treated individuals (BPD), the weighted average of reinfections is %s (%s-%s), for %s recurrences',
                   round(100*sum(MS_final$I_mean[ind_PMQ_BPD],na.rm=T)/length(ind_PMQ_BPD),1),
                   round(100*sum(MS_final$I_lower[ind_PMQ_BPD],na.rm=T)/length(ind_PMQ_BPD),1),
                   round(100*sum(MS_final$I_upper[ind_PMQ_BPD],na.rm=T)/length(ind_PMQ_BPD),1),
                   length(ind_PMQ_BPD)))

```

# False positive rate of relapse

We want to know how often our model estimates evidence of relapse across pairs of episodes when the episodes are in different people (i.e. have not possibility of being a relapse)

```{r}
if(RUN_MODELS_FALSE_POSITIVE){
  # check if the massive pairwise dataset has been made, if not make it 
  # (takes a long time ~20hours)
  if(!"APC_MSdata.bigRData"%in%list.files()){
    # The pooled MS data from BPD and VHX
    load('../RData/GeneticModel/MS_data_PooledAnalysis.RData')
    tic()
    APC_MSdata = Make_All_Pairwise_Comparisons(MS_data = MS_pooled, ncores=42)
    save(APC_MSdata, file = 'APC_MSdata.bigRData')
    toc()
  } 
  load('APC_MSdata.bigRData')
  print('The inflated pairwise dataset is available, now running the analysis...')
  # Run the genetic model on the pairwise data
  tic()
  Inflated_Results = post_prob_CLI(MSdata = APC_MSdata, 
                                   Fs = Fs_Combined, 
                                   UpperComplexity = 10^6, 
                                   verbose = F,
                                   cores = 42)
  toc()
  save(Inflated_Results, file = 'Inflated_Results.bigRData')
} else {
  load('~/Dropbox/RecurrentVivax/Pooled_Final_Analysis/Inflated_Results.bigRData')
  Inflated_Results = Inflated_Results[!is.na(Inflated_Results$L),]
  load('~/Dropbox/RecurrentVivax/Pooled_Final_Analysis/APC_MSdata.bigRData')
}
```


```{r, echo=FALSE}
writeLines(sprintf('The false-positive discovery rate of the genetic model is estimated as %s percent. 
                   \nThis is based on %s pairwise comparisons',
                   round(100*sum( Inflated_Results$L > Epsilon_upper )/nrow(Inflated_Results),2),
                   nrow(Inflated_Results)))
```

# Analysis of radical cure efficacy in BPD

Almost all episodes in BPD were typed. Therefore we can estimate the true efficacy comparing with historical controls (VHX).

```{r}
Combined_Time_Data$Episode_Identifier = apply(Combined_Time_Data,1,
                                              function(x){
                                                paste(x['patientid'],as.integer(x['episode']),
                                                      sep='_')} )

# iterate over every episode and use either the joint posterior 
# or if missing the time probability (this could be time censored probability)
Combined_Time_Data$Reinfection_Probability=
  Combined_Time_Data$Reinfection_Probability_LL=
  Combined_Time_Data$Reinfection_Probability_UL = NA

Mod2_ThetaEstimates$Failure_Identifier = 
  apply(Mod2_ThetaEstimates, 1, 
        function(x) paste(x['patientid'], as.integer(x['episode'])-1,sep='_'))

sss=0
for(i in 1:nrow(Combined_Time_Data)){
  ep_id = Combined_Time_Data$Episode_Identifier[i]
  MS_id = paste(Combined_Time_Data$patientid[i],
                as.integer(Combined_Time_Data$episode[i])+1, sep='_')
  # If in MS_final then use full probability
  if(MS_id %in% MS_final$Episode_Identifier){
    Combined_Time_Data$Reinfection_Probability[i] =
      MS_final$I_mean[MS_final$Episode_Identifier==MS_id]
    Combined_Time_Data$Reinfection_Probability_UL[i] =
      MS_final$I_upper[MS_final$Episode_Identifier==MS_id]
    Combined_Time_Data$Reinfection_Probability_LL[i] =
      MS_final$I_lower[MS_final$Episode_Identifier==MS_id]
  } else { # use the time to event model
    ind = which(Mod2_ThetaEstimates$Failure_Identifier==ep_id)
    if(length(ind)>0){
      Combined_Time_Data$Reinfection_Probability[i] =
        Mod2_ThetaEstimates$ReInfection_mean_theta[ind]
      Combined_Time_Data$Reinfection_Probability_UL[i] =
        Mod2_ThetaEstimates$ReInfection_975_theta[ind]
      Combined_Time_Data$Reinfection_Probability_LL[i] =
        Mod2_ThetaEstimates$ReInfection_025_theta[ind]
      sss=sss+1
    }
  }
}
```


Now we look at whether the PK (carboxy-primaquine) can predict failure:
First we add the carboxy to the dataset:
```{r}
Combined_Time_Data = arrange(Combined_Time_Data, patientid, episode)
load('../RData/PK_data/BPD_pk.RData')
BPD_pk = filter(BPD_pk, !is.na(Episode))
Combined_Time_Data$log10_carboxyPMQ = NA
Combined_Time_Data$log10_PMQ = NA
Combined_Time_Data$NumberDaysPMQ = 14
for(i in 1:nrow(Combined_Time_Data)){
  id = Combined_Time_Data$patientid[i]
  ep_i = Combined_Time_Data$episode[i]
  all_id_eps = Combined_Time_Data$episode[Combined_Time_Data$patientid==id]
  pk_ind = which(BPD_pk$ID == id & BPD_pk$Episode==ep_i)
  if(length(pk_ind)>0){
    if(length(pk_ind)>1) print(id)
    Combined_Time_Data$log10_carboxyPMQ[i] = mean(BPD_pk$log10_carboxyPQ_PK[pk_ind])
    Combined_Time_Data$log10_PMQ[i] = mean(BPD_pk$log10_PQ_PK[pk_ind])
    Combined_Time_Data$NumberDaysPMQ[i] = BPD_pk$NumberofPKDays[pk_ind[1]]
  }
}
```

We exclude the two recurrences seen in patient BPD_444

```{r CarboxyPredictionFailure}
# These are two outliers - have discussed with Cindy
BPD444_recurrences = Combined_Time_Data$patientid=='BPD_44' & Combined_Time_Data$episode>1
BPD_598 = which(Combined_Time_Data$patientid=='BPD_598')
ind_keep = !BPD444_recurrences #& !BPD_598
require(lme4)
Combined_Time_Data$Failure_YN = Combined_Time_Data$Reinfection_Probability < 0.5
mod = glmer(Failure_YN ~ log10_carboxyPMQ + NumberDaysPMQ + 
              (1 | patientid), 
            family = 'binomial', data=Combined_Time_Data[ind_keep,])
summary(mod)
# Plot the data and model
xs=seq(0,4,by=.01)
par(las = 1, bty='n')
regimen_colors = brewer.pal(8, 'Dark2')[c(1,5)]
plot(Combined_Time_Data$log10_carboxyPMQ[ind_keep]*Combined_Time_Data$NumberDaysPMQ[ind_keep],
     jitter(as.numeric(Combined_Time_Data$Failure_YN[ind_keep]), factor = 0.02), 
     col = regimen_colors[mapvalues(Combined_Time_Data$NumberDaysPMQ[ind_keep],c(7,14),c(1,2))],
     pch = mapvalues(Combined_Time_Data$NumberDaysPMQ[ind_keep],c(7,14),c(3,4)),
     xlab = 'Carboxy-primaquine exposure: days * log(ng/mL)',
     panel.first = grid(), ylab = 'Probability of failure')
legend(x = 40, y = 0.3, bty='n', col =regimen_colors, 
       pch=c(3,4), legend = c('7 days','14 days'))
lines(xs*7, predict(mod, newdata=data.frame(log10_carboxyPMQ=xs, NumberDaysPMQ=7,
                                            patientid='new'),allow.new.levels=T,
                    type='response'), lwd=2, col= regimen_colors[1], lty=2)
lines(xs*14, predict(mod, newdata=data.frame(log10_carboxyPMQ=xs, NumberDaysPMQ=14,
                                             patientid='new'),allow.new.levels=T,
                     type='response'), lwd=2, col= regimen_colors[2], lty = 2)
points(Combined_Time_Data$log10_carboxyPMQ[BPD_598]*Combined_Time_Data$NumberDaysPMQ[BPD_598],
       Combined_Time_Data$Failure_YN[BPD_598], cex=2)
text(Combined_Time_Data$log10_carboxyPMQ[BPD_598[1]]*
       Combined_Time_Data$NumberDaysPMQ[BPD_598[1]]+3,
     Combined_Time_Data$Failure_YN[BPD_598[1]]-0.05, labels = 'BPD_598')
```

Now we remove outliers and fit the same model (CPMQ outliers)
```{r}
mu_hat_7 = mean(Combined_Time_Data$log10_carboxyPMQ[Combined_Time_Data$NumberDaysPMQ==7])
sd_hat_7 = sd(Combined_Time_Data$log10_carboxyPMQ[Combined_Time_Data$NumberDaysPMQ==7])
mu_hat_14 = mean(Combined_Time_Data$log10_carboxyPMQ[Combined_Time_Data$NumberDaysPMQ==14],na.rm=T)
sd_hat_14 = sd(Combined_Time_Data$log10_carboxyPMQ[Combined_Time_Data$NumberDaysPMQ==14],na.rm=T)

par(mfrow=c(1,2))
hist(Combined_Time_Data$log10_carboxyPMQ[Combined_Time_Data$NumberDaysPMQ==7],
     main='', xlab='carboxy primaquine day 7 (ng/mL)')
abline(v = mu_hat_7 - sd_hat_7*3)

hist(Combined_Time_Data$log10_carboxyPMQ[Combined_Time_Data$NumberDaysPMQ==14],
     main = '', xlab='carboxy primaquine day 7 (ng/mL)')
abline(v = mu_hat_14 - sd_hat_14*3)

outliers7 = Combined_Time_Data$NumberDaysPMQ==7 & Combined_Time_Data$log10_carboxyPMQ<mu_hat_7 - sd_hat_7*3
outliers14 = Combined_Time_Data$NumberDaysPMQ==14 & Combined_Time_Data$log10_carboxyPMQ<mu_hat_14 - sd_hat_14*3

mod_No_Outliers = glmer(Failure_YN ~ log10_carboxyPMQ + NumberDaysPMQ + 
                          (1 | patientid), 
                        family = 'binomial', data=Combined_Time_Data[ind_keep & !outliers14 & !outliers7,])
summary(mod_No_Outliers)
```

Compare results with and without outliers:

```{r CarboxyPredictionFailure_NoOutliers}
par(las = 1, bty='n')
plot(Combined_Time_Data$log10_carboxyPMQ[ind_keep]*Combined_Time_Data$NumberDaysPMQ[ind_keep],
     jitter(as.numeric(Combined_Time_Data$Failure_YN[ind_keep]),factor = 0.03), 
     col = regimen_colors[mapvalues(Combined_Time_Data$NumberDaysPMQ[ind_keep],c(7,14),1:2)],
     pch = mapvalues(Combined_Time_Data$NumberDaysPMQ[ind_keep],c(7,14),c(3,4)),
     xlab = 'Carboxy-primaquine trough exposure: days * log(ng/mL)',
     ylab = 'Probability of failure', panel.first = grid())
legend(x = 25, y = 0.7, bty='n', col = c(1, regimen_colors, 1,1),
       pch=c(1,3,4,NA,NA), lty = c(NA,NA,NA,1,2),lwd=c(NA,NA,NA,2,2),
       legend = c('Outlier','7 days','14 days',
                  'Regression fit: all data', 'Regression fit: outliers removed'))
lines(xs*7, predict(mod, newdata=data.frame(log10_carboxyPMQ=xs, NumberDaysPMQ=7,
                                            patientid='new'),allow.new.levels=T,
                    type='response'), lwd=2, col= regimen_colors[1], lty=1)
lines(xs*14, predict(mod, newdata=data.frame(log10_carboxyPMQ=xs, NumberDaysPMQ=14,
                                             patientid='new'),allow.new.levels=T,
                     type='response'), lwd=2, col= regimen_colors[2], lty = 1)

lines(xs*7, predict(mod_No_Outliers, newdata=data.frame(log10_carboxyPMQ=xs, NumberDaysPMQ=7,
                                                        patientid='new'),allow.new.levels=T,
                    type='response'), lwd=2, col= regimen_colors[1], lty=2)
lines(xs*14, predict(mod_No_Outliers, newdata=data.frame(log10_carboxyPMQ=xs, NumberDaysPMQ=14,
                                                         patientid='new'),allow.new.levels=T,
                     type='response'), lwd=2, col= regimen_colors[2], lty = 2)

# outline the outliers
points(Combined_Time_Data$log10_carboxyPMQ[outliers14|outliers7]*Combined_Time_Data$NumberDaysPMQ[outliers14|outliers7],
       Combined_Time_Data$Failure_YN[outliers14|outliers7], cex=2)
```

Now we calculate a compressed dataset and failure for each individual

```{r}
# now we calculate the primaquine failure rate
# For individuals with two episodes: P(failure) = 1 - P(Rec 1 = I)*P(Rec 2 = I)
Summary_data = Combined_Time_Data[!duplicated(Combined_Time_Data$patientid),]
Summary_data$Failure_UL = Summary_data$Failure_LL = 
  Summary_data$Failure = Summary_data$CPMQ = 
  Summary_data$CPMQ = NA
for(i in 1:nrow(Summary_data)){
  ind = which(Combined_Time_Data$patientid==Summary_data$patientid[i])
  Summary_data$Failure[i] = 1-prod(Combined_Time_Data$Reinfection_Probability[ind],na.rm=T)
  Summary_data$Failure_UL[i] = 1-prod(Combined_Time_Data$Reinfection_Probability_UL[ind],na.rm=T)
  Summary_data$Failure_LL[i] = 1-prod(Combined_Time_Data$Reinfection_Probability_LL[ind],na.rm=T)
  Summary_data$CPMQ[i] = median(Combined_Time_Data$log10_carboxyPMQ[ind],na.rm=T)
}
BPD_data = Summary_data[grep('BPD', Summary_data$patientid),]

P_Failure=100*sum(BPD_data$Failure)/nrow(BPD_data)
# invert the intervals here - optimistic for not failure = pessimistic for failure
P_Failure_UL = 100*sum(BPD_data$Failure_LL)/nrow(BPD_data)
P_Failure_LL = 100*sum(BPD_data$Failure_UL)/nrow(BPD_data)

writeLines(sprintf('The primaquine failure rate in the %s individuals is %s%% (%s-%s) over the course of %s years total follow-up.',
                   nrow(BPD_data), round(P_Failure,2),
                   round(P_Failure_LL,2),
                   round(P_Failure_UL,2), round(sum(BPD_data$FU_time)/365)))
```


This won't go into this paper but looking out of interest:
Does 2D6 correlate with carboxy ?

```{r}
TwoD6_dat = read.csv('~/Dropbox/RecurrentVivax/RData/PK_data/TwoD6&Vivax Genotyping_ASscore.csv')
TwoD6_dat$ID = apply(TwoD6_dat, 1, function(x) paste(x['Study'],
                                                     as.integer(x['Patient.ID']),
                                                     sep = '_'))
TwoD6_dat$Phenotype = mapvalues(TwoD6_dat$X2D6.Phenotype, 
                                from = c('PM','IM', 'EM'), to = 1:3)
Combined_Time_Data$Phenotype = Combined_Time_Data$ASscore = NA
for(i in 1:nrow(Combined_Time_Data)){
  id = Combined_Time_Data$patientid[i]
  if(sum(TwoD6_dat$ID==id)>0){
    Combined_Time_Data$ASscore[i] = TwoD6_dat$AS.score[TwoD6_dat$ID==id]
    Combined_Time_Data$Phenotype[i] = TwoD6_dat$Phenotype[TwoD6_dat$ID==id]
  }
}

mod_2D6 = lmer(log10_carboxyPMQ ~ ASscore + NumberDaysPMQ + (1 | patientid), 
               data = Combined_Time_Data)
summary(mod_2D6)
plot(Combined_Time_Data$ASscore, Combined_Time_Data$log10_carboxyPMQ, pch=20)
lines(xs, predict(mod_2D6, data.frame(ASscore=xs,NumberDaysPMQ=7,patientid='new'),allow.new.levels=T),lwd=3)
```


```{r}
Combined_2D6data = filter(Combined_Time_Data, !is.na(ASscore), !Censored)
for(id in unique(Combined_2D6data$patientid)){
  ind = Combined_2D6data$patientid==id
  Combined_2D6data$Failure_YN[ind] = max(Combined_2D6data$Failure_YN[ind])
}
Combined_2D6data = Combined_2D6data[!duplicated(Combined_2D6data$patientid),]
mod_Failure = glm(Failure_YN ~ ASscore, 
                  data = Combined_2D6data, family = 'binomial')
summary(mod_Failure)
```


# Extra Analyses

## Looking at the effect of inbreeding coefficient

Our model has a parameter $\alpha$ which defines the level of inbreeding within the population.
Taylor is developing methods for the estimation of $\alpha$ from genetic data (in preparation).

We look at the sensitivity of the results (all the above is with $\alpha=0$) for a reasonable upper bound of $\alpha=0.175$.

We rerun the analysis on the single run isolates (low computational complexity):

```{r}
alphaUpper = 0.175
if(RUN_MODELS_SINGLE_SIMPLE){
  #===============================================
  # Run (with time-to-event)
  #===============================================
  # Approx 100 secs per full model run
  tic()
  thetas_9MS_alphaUpper = post_prob_CLI(MSdata = MS_pooled, Fs = Fs_Combined, 
                                        p = p, cores = 6, verbose = F, alpha = alphaUpper) 
  thetas_9MS_alphaUpper$Episode_Identifier = rownames(thetas_9MS_alphaUpper)
  save(thetas_9MS_alphaUpper, file = '../RData/GeneticModel/thetas_9MS_alphaUpper.RData')
  toc()
} else {
  load('../RData/GeneticModel/thetas_9MS_alphaUpper.RData')
}
```

```{r}
par(las=1, bty='n', mfrow=c(2,2))
plot(log10(thetas_9MS$L), log10(thetas_9MS_alphaUpper$L),
     ylab = 'Relapse: inbred',
     xlab = 'Relapse: no inbreeding',
     col= drug_cols2[thetas_9MS$drug],pch=20)
lines(c(-10,10),c(-10,10),lty=2)

plot(log10(thetas_9MS$L), log10(thetas_9MS$L)-log10(thetas_9MS_alphaUpper$L),
     ylab = 'Difference in Log Relapse Probabilities',
     xlab = 'Relapse Probability: no inbreeding',
     col= drug_cols2[thetas_9MS$drug],pch=20)
abline(h=0,lty=2)
legend('topleft', legend = c('No PMQ', 'PMQ+'), col = drug_cols2[2:3], pch = 20, bty = 'n')

###****** Reinfection : comparison *****#####
par(las=1, bty='n')
plot(log10(thetas_9MS$I), log10(thetas_9MS_alphaUpper$I),
     ylab = 'Reinfection: inbred',
     xlab = 'Reinfection: no inbreeding',
     col= drug_cols2[thetas_9MS$drug],pch=20)
lines(c(-10,10),c(-10,10),lty=2)

plot(log10(thetas_9MS$I), log10(thetas_9MS$I)-log10(thetas_9MS_alphaUpper$I),
     ylab = 'Difference in Log Reinfection Probabilities',
     xlab = 'Reinfection: no inbreeding',
     col= drug_cols2[thetas_9MS$drug],pch=20)
legend('topleft', legend = c('No PMQ', 'PMQ+'), col = drug_cols2[2:3], pch = 20, bty = 'n')

abline(h=0,lty=2)
```

Interpretation: Adding the inbreeding coefficent slightly changes some of the probabilities of relapse for some primaquine treated individuals (only green dots are being shifted).
This means that inbreeding would imply that fewer of the primaquine treated episodes are relapses, implying higher efficacy of the drug.

For the non-primaquine group, it is just tempering the very low probabilities of reinfection seen for some episodes.

In conclusion, this isn't changing the results significantly and would imply a greater primaquine efficacy that reported in the paper.
